### ISSUE TITLE
Strict null check control flow bug

### ISSUE THREAD

**Issue by ConradBrean** [Microsoft/TypeScript#25575]
**Strict, null, union type, control flow, TS2339** 

**Code**

```ts
interface T1{
  test1: number;
}

interface T2{
  test2: T1;
}

class TestClass{
  field: T2 | null = null;

  mutateField(): void{
    if (Math.random() > 0.5){
      this.field = {
        test2: {
          test1: 1
        }
      }
    } else {
      this.field = null;
    }
  }

}

function testFunc(){
  const classInstance = new TestClass();
  classInstance.field = null;

  classInstance.mutateField();

  const field = classInstance.field;

  if (field !== null){
    console.log(field.test1);
  }
}

```

**Expected behavior:** No error.

**Actual behavior:** 
Error: nullControlFlowCheck.ts(35,23): error TS2339: Property 'test1' does not exist on type 'never'.

Compiler infers that  classInstance.field is null even so it have been mutated after null assignment

[Playground](https://agentcooper.github.io/typescript-play/?noUnusedLocals=true&noUnusedParameters=true&preserveConstEnums=true#code/JYOwLgpgTgZghgYwgAgCoEYDeAoZzIDOY6AXMiAK4C2ARtANzYC+22oksiKqATDnoTA8yGRi2wIANnAIE0EIgGFps-shjAIkgCYieyAD7kKkycgC8x041zIqFMHEgAxTToAUASjIA3APbA2mp4wDDI7gCyTgAWAHRQcCDaflReyAB8yAAMsQCsnsF4+NHABLEaWtoWyIVF+ApCZLV19USkyOi2Lcgs3b1FTMhaBCjNYCVlFTrVlNZdPbbi4jAUIAhgwH4grWDOqwheaghbRMhSMgQAkiBEiUgzEADu8koqBF6MeOey17drEOU3FVLLNJDYvm9fo5-rF7I4XECPqwvicwOogdVvlcbtCkIDKuDkKFwlMqgBCcwgkySArzY43PySAGSPwAc3cpNignQnk+CxYQA)

**Comment by ajafff** [comment]
Duplicate of #9998
